taxjar_salestax/models/account_invoice.py

```diff
@@ -0,0 +1,285 @@
+from odoo.addons.queue_job.job import job
+from odoo.exceptions import UserError
+from odoo.tools.translate import _
+
+
+class AccountInvoice(models.Model):
+    _inherit = 'account.invoice'
+
+    warehouse_id = fields.Many2one(
+        'stock.warehouse', 'Warehouse',
+        help=('Determines the TaxJar origin address. You can override it '
+              'here without affecting any related stock operations.'))
+    taxjar_applies = fields.Boolean(compute="_compute_taxjar_applies")
+    taxjar_amount = fields.Float(
+        string='TaxJar Amount', digits=dp.get_precision('Account'),
+        readonly=True)
+    taxjar_shipping_amount = fields.Float(
+        string='Shipping Cost', digits=dp.get_precision('Account'),
+        store=True, readonly=True, compute='_compute_taxjar_shipping_amount')
+
+    def _compute_taxjar_applies(self):
+        """
+        Determine if the invoice's related company has a Taxjar
+        configuration that is valid for this invoice.
+        """
+        for invoice in self:
+            invoice.taxjar_applies = (
+                invoice.type in ('out_invoice', 'out_refund') and bool(
+                    invoice._get_taxjar_params()[0]))
+
+    @api.depends('invoice_line_ids.price_subtotal')
+    def _compute_taxjar_shipping_amount(self):
+        for inv in self:
+            config, partner = inv._get_taxjar_params()
+            if not config:
+                return
+            shipping_amount = 0.0
+            for line in inv.invoice_line_ids:
+                shipping_amount += line._get_taxjar_shipping_amount(config)
+            sign = -1 if inv.type == 'out_refund' else 1
+            inv.taxjar_shipping_amount = sign * shipping_amount
+
+    def _get_taxjar_partner(self):
+        """ Retrieve the delivery address, for tax purposes """
+        self.ensure_one()
+        order = self.get_sale_order()
+        if order:
+            return order._get_taxjar_partner()
+        return self.partner_id
+
+    def get_sale_order(self):
+        """ Return the first sale order linked to this invoice """
+        self.ensure_one()
+        order_lines = (
+            self.mapped('invoice_line_ids.sale_line_ids') or
+            self.refund_invoice_id.mapped('invoice_line_ids.sale_line_ids'))
+        if order_lines:
+            return order_lines[0].order_id
+        if self.origin:  # support legacy orders
+            a = self.origin
+            if len(a.split(':')) > 1:
+                so_origin = a.split(':')[1]
+            else:
+                so_origin = a.split(':')[0]
+            return self.env['sale.order'].search(
+                [('name', '=', so_origin)], limit=1)
+        return self.env['sale.order']
+
+    def _get_taxjar_origin_address(self):
+        """ partner address, on which TaxJar tax will calculate  """
+        self.ensure_one()
+        order = self.get_sale_order()
+        if order and order.warehouse_id and order.warehouse_id.partner_id:
```

Comment:
this is very V8y - why not

```suggestion
        if order.warehouse_id.partner_id:
```